core/issuers.chainmail

```diff
@@ -1,56 +1,174 @@
-  getLabel() ::{ issuer ::Issuer, description };
-  getAssay() ::Assay;
-  makeEmptyPurse(name ?String) ::Purse;
-
-  class Mint {
-    getIssuer() Issuer;
-    mint(initialBalance ?Amount, name ?String) ::Purse;
-  }
-  class Payment {
-    getIssuer() ::Issuer;
-    getXferBalance() ::Amount;
-  }
-  class Purse {
-    getIssuer() ::Issuer;
-    getXferBalance() ::Amount;
-    getUseBalance() ::Amount;
-    deposit(amount ?Amount, srcPaymentP ?reveal[Promise]) ::Amount;
-    withdraw(amount ?Amount, name ?String) ::Purse;
-  }
+
+/**
+ * Payments hold verified amounts of certain rights issued by Mints. Amounts
+ * from payments can be deposited in purses, but otherwise, the entire amount is
+ * available when the payment is transferred. A payment's balance can only fall,
+ * through the action of withdraw.
+ * 
+ * Payments can be converted to Purses by calling
+ * payment.getIssuer().makeEmptyPurse().depositAll(payment);
+ */
+interface Payment extends (PurseOrPayment) {
+  getIssuer() -> (Issuer);
+  getXferBalance() -> (Amount);
+}
+
+/**
+ * PurseOrPayment is the parent type of Purse and Payment. Most interfaces only
+ * accept one of the sub-types to reduce the incidence of the common mistake of
+ * giving away a purse when you meant to hold on to the purse and give away part
+ * of the contents. deposit() and depositAll() are the two exceptions, for which
+ * the caller explicitly says to give up all contained rights, or provides an
+ * explicit amount to transfer.
+ */
+interface PurseOrPayment {
+  getIssuer() -> (Issuer);
+  getXferBalance() -> (Amount);
 }
-makeMint(description, makeAssay ::(Label -> Assay)) ::Issuer[Assay].Mint;
-
-class Peg[RemoteIssuer, LocalIssuer] {
-  getLocalIssuer() ::LocalIssuer;
-  getRemoteIssuer() ::reveal<RemoteIssuer>;
-  retain(remoteAmount ?reveal<RemoteIssuer.Amount>,
-         remotePaymentP ?reveal<RemoteIssuer.Payment>,
-         name ?String) ::LocalIssuer.Payment;
-  redeem(localAmount ?LocalIssuer.Amount,
-         localPayment ?LocalIssuer.Payment,
-         name ?String) ::reveal<RemoteIssuer.Payment>;
+
+/** MintController records the amounts contained by purses. */
+interface MintController {
+  /** (Non-fungible amounts only.) Ensure no purse contains amount. */
+  destroy(amount:Amount);
+
+  /** Reset all purses to contain nothing. */
+  destroyAll();
+
+  /**
+   * Record that payment is a new Payment containing amount, and that src
+   * now contains srcNewRightsAmount.
+   */
+  recordPayment(src, payment, amount, srcNewRightsAmount);
+
+  /**
+   * Record that srcPayment now contains srcNewRightsAmount, and purse is a new
+   * Purse containing purseNewRightsAmount.
+   */
+  recordDeposit(scPayment, srcNewRightsAmount, purse, purseNewRightsAmount);
+
+  /**
+   * Record that purse is a new Purse containing srcNewRightsAmount.
+   */
+  recordMint(purse:Purse, initialAmount:Amount);
+
+  /** Return the Amount contained in purse. */
+  getAmount(purse:Purse) -> (Amount);
```

Comment:
purse or payment

Reply:
Yeah, these will all have to change with your cleanup.